urlencode() new query params values

fgilio opened the pull request:
Hi guys 👋🏼
I'm making this PR in part to signal the bug, because maybe there's a better solution.
We've had a bunch of reports from links not working correctly and nailed it down to the query parameters not being URL encoded.
- Fails: https://mysite.com/filter?category=cute cats
- Works: https://mysite.com/filter?category=cute%20cats

Of course, it's the same for `&`, `?`, etc

So basically I just added `urlencode()` to `withQueryParameter()`.

What do you think? Ever had this issue?

freekmurze commented:
This seems like a good chance. It's breaking so we'll have to release a new major version for this which I'm willing to do.

I think url encoding is better dan in the query parameter bag class. Could you also add a test to make sure that this works?

fgilio commented:
👋🏼 I worked a little bit more on this:

- Switched to `rawurlencode` as it follows a standard RFC: https://www.php.net/manual/en/function.rawurlencode.php
    - The only practical difference is that it'll correctly encode spaces to `%20`, instead of `+` as does `urlencode`
- Also tweaked `QueryParameterBag::fromString` to accept url encoded values
- Added tests for both `QueryParameterBag::fromString` and `__toString` to make sure everything works alright, all test pass!

Let me know what you think

brendt commented:
Seems good to me! @freekmurze I not entirely familiar with this package, what exactly would result in a breaking change?

freekmurze commented:
I saw it as a breaking change because the output now changes, but we could indeed view this as a bug fix release too.

freekmurze commented:
Thanks!

fgilio commented:
🎉 thanks guys!

NBZ4live commented:
It was released as patch version and not as major version which broke everywhere where urlencode was used manually

fgilio commented:
@brendt should this be rolled back and released again as a 2.0?

brendt commented:
@freekmurze we might need to release it as a v2 instead if people are indeed experiencing it as an issue.

fgilio commented:
I could try to detect if the URL is already encoded by checking if there's any `+` or `%` signs (or other better way) and avoid re encoding.
A v2 would be the safest way, but I can try that if you want.
